README.md
# multiselect_scope

Package for multiple selection and tracking selected items (select all, clear, invert selection, etc.)

## Getting Started

![](.assets/screencast_small.gif)

## Usage

1. Create `MultiselectController` if you need to manage selection outside `MultiselectScope`. 
```dart
final multiselectController = MultiselectController();
```

2. Wrap your multichild widget (`ListView` or `GridView` or another) in `MultiselectScope`. Pass to `dataSource` list of domain objects (`Cars` or `Employes` for example). Do not forget pass domain type to generic parameter (`MultiselectScope<String>` if domain type is `String`). Now let's look at the comments in the code 
```dart
MultiselectScope<String>(
  controller: multiselectController,
  dataSource: items,
  // Set this to true if you want automatically 
  // clear selection when user tap back button
  clearSelectionOnPop: true,
  // When you update [dataSource] then selected indexes will update
  // so that the same elements in new [dataSource] are selected
  keepSelectedItemsBetweenUpdates: true,
  initialSelectedIndexes: [1, 3],
  // Callback that call on selection changing
  onSelectionChanged: (indexes, items) {

  },
  child: ListView.builder(
      itemCount: items.length,
      itemBuilder: (context, index) {
        // Getting link to controller  
        final controller = MultiselectScope.controllerOf(context);

        final itemIsSelected = controller.isSelected(index);

        return InkWell(
          // You may implement selection logic as you want
          // For example start selection mode on long tap  
          onLongPress: () {
            if (!controller.selectionAttached) {
              controller.select(index);
            }
          },
          onTap: () {
            if (controller.selectionAttached) {
              controller.select(index);
            }
          },
          child: Padding(
            padding: const EdgeInsets.all(8.0),
            child: Container(
              // Colorize item depend on selected it or not 
              color: itemIsSelected ? Theme.of(context).primaryColor : null,
              child: Text(
                items[index],
                style: TextStyle(fontSize: 22),
              ),
            ),
          ),
        );
      }),
)
      

```

3. Control you selection

Get link to your controller inside scope:
```dart
final controller = MultiselectScope.controllerOf(context);
```
if your want to use controller outside scope that get link that you create before (`multiselectController`)

Manage selection:
```dart
multiselectController.selectAll();
multiselectController.select(0);
multiselectController.invertSelection();
multiselectController.clearSelection();
multiselectController.setSelectedIndexes([1,2,3]);
 
```

4. Get selected items
```dart
final selectedItems = multiselectController.getSelectedItems().cast<String>();
```  

or check item is selected:

```dart
final itemIsSelected = controller.isSelected(index);
```                          
 


CHANGELOG.md
## [0.2.0] - Update to null safety

## [0.1.1] - Fix example

## [0.1.0] - Release

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import '../lib/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}

test/multiselect_scope_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:multiselect_scope/multiselect_scope.dart';
import 'package:collection/collection.dart';

extension on List? {
  bool containsAll(Iterable items) {
    for (var element in items) {
      if (!this!.contains(element)) {
        return false;
      }
    }

    return true;
  }

  bool notContainsAll(Iterable items) {
    for (var element in items) {
      if (this!.contains(element)) {
        return false;
      }
    }

    return true;
  }
}

void main() {
  List<String>? _stateItems;

  void _updateStateItems(_MyHomePageTestState stateBy) {
    _stateItems =
        stateBy.multiselectController.getSelectedItems().cast<String>();
  }

  testWidgets('Multiselect test', (WidgetTester tester) async {
    await tester.pumpWidget(MyAppTest());

    final state =
        tester.state(find.byType(MyHomePageTest)) as _MyHomePageTestState;
    debugPrint(state.items.toString());

    // 1. Ensure that two starting elements was selected
    _updateStateItems(state);
    expect(_stateItems!.contains('Item 1'), true);
    expect(_stateItems!.contains('Item 3'), true);

    // 2. Click on the 7th element, make sure it is selected
    await tester.tap(find.text('Item 7'));
    await tester.pump();
    _updateStateItems(state);

    expect(_stateItems!.contains('Item 7'), true);
    expect(state.selectedItemsTrack.containsAll(['Item 7', 'Item 1', 'Item 3']),
        true);

    // 3. Кликнуть на кнопку Invert , убедиться, что другие элементы выделены
    await tester.tap(find.text('Invert'));
    await tester.pump();
    _updateStateItems(state);

    expect(
        _stateItems.containsAll(
            state.items.toSet().difference({'Item 7', 'Item 1', 'Item 3'})),
        true);
    expect(state.selectedIndexesTrack.containsAll([0, 2, 5]), true);
    expect(state.selectedIndexesTrack.notContainsAll([7, 1, 3]), true);

    // 4. Кликнуть на кнопку Select all, убедиться, что все элементы выделены
    await tester.tap(find.text('Select all'));
    await tester.pump();
    _updateStateItems(state);

    expect(_stateItems.containsAll(state.items.toSet()), true);

    // 5. Инвертировать (теперь список элементов пуст)
    await tester.tap(find.text('Invert'));
    await tester.pump();
    _updateStateItems(state);

    expect(
        _stateItems!.isEmpty &&
            state.selectedIndexesTrack!.isEmpty &&
            state.selectedItemsTrack!.isEmpty,
        true);

    // 6. Выделить 4 и 7 элементы нажать на кнопку Add rand, Убедиться, что остались выделены те же эелементы
    await tester.longPress(find.text('Item 4'));
    await tester.pump();
    await tester.tap(find.text('Item 7'));
    await tester.pump();

    await tester.tap(find.text('Add rand'));
    await tester.pump();

    _updateStateItems(state);

    final preselected = ['Item 4', 'Item 7'];
    expect(state.selectedItemsTrack.containsAll(preselected), true);
    expect(_stateItems.containsAll(preselected), true);

    // 7.  Нажать кнопку Remove rand, убедиться, что остались выделены те же элементы (если один из них не выделен, то возможно удален)
    await tester.tap(find.text('Remove rand'));
    await tester.pump();
    _updateStateItems(state);

    preselected.forEach((String preselectedItem) {
      if (state.items.contains(preselectedItem)) {
        expect(_stateItems?.contains(preselectedItem), true);
      }
      ;
    });

    // 8. Нажать кнопку Clear, убедиться, что список пуст
    await tester.tap(find.text('Clear'));
    await tester.pump();
    _updateStateItems(state);

    expect(
        _stateItems!.isEmpty &&
            state.selectedIndexesTrack!.isEmpty &&
            state.selectedItemsTrack!.isEmpty,
        true);

    // 9. Выделить 3 элемент, нажать кнопку Delete, убедиться, что он удалился
    await tester.longPress(find.text('Item 3'));
    await tester.pump();
    _updateStateItems(state);

    await tester.tap(find.text('Delete'));
    await tester.pump();
    _updateStateItems(state);

    expect(state.items.contains('Item 3'), false);

    final stateIndexes = List.from(state.selectedIndexesTrack!);

    await tester.tap(find.text('No keep'));
    await tester.pump();

    await tester.tap(find.text('Add rand'));
    await tester.pump();

    _updateStateItems(state);

    Function eq = const ListEquality().equals;
    expect(eq(stateIndexes, state.selectedIndexesTrack), true);

    // Test specified index selection
    await tester.tap(find.text('Select specified'));
    await tester.pump();
    _updateStateItems(state);

    expect(eq([0, 1, 2], state.selectedIndexesTrack), true);
  });
}

class MyAppTest extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Multiselect scope Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePageTest(title: 'Multiselect'),
    );
  }
}

class MyHomePageTest extends StatefulWidget {
  MyHomePageTest({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageTestState createState() => _MyHomePageTestState();
}

class _MyHomePageTestState extends State<MyHomePageTest> {
  late List<String> items;
  List<String>? selectedItemsTrack;
  List<int>? selectedIndexesTrack;
  late bool keepSelectedIndexes;

  late MultiselectController multiselectController;
  late Random random;

  @override
  void initState() {
    super.initState();
    keepSelectedIndexes = true;
    random = Random();
    items = List.generate(10, (index) => 'Item $index');

    multiselectController = MultiselectController();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: MultiselectScope<String>(
        controller: multiselectController,
        dataSource: items,
        clearSelectionOnPop: true,
        initialSelectedIndexes: [1, 3],
        keepSelectedItemsBetweenUpdates: keepSelectedIndexes,
        onSelectionChanged: (indexes, items) {
          selectedItemsTrack = items;
          selectedIndexesTrack = indexes;

          debugPrint(
              'Custom listener invoked! Indexes: $indexes Items: $items');
          return;
        },
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            children: <Widget>[
              Expanded(
                child: ListView.builder(
                    itemCount: items.length,
                    itemBuilder: (context, index) {
                      final controller = MultiselectScope.controllerOf(
                          context); //MultiselectScope.of(context);

                      final itemIsSelected = controller.isSelected(index);

                      return InkWell(
                        onLongPress: () {
                          if (!controller.selectionAttached) {
                            controller.select(index);
                          }
                        },
                        onTap: () {
                          debugPrint('Item is selected: $itemIsSelected');

                          if (controller.selectionAttached) {
                            controller.select(index);
                          }
                        },
                        child: Padding(
                          padding: const EdgeInsets.all(8.0),
                          child: Container(
                            color: itemIsSelected
                                ? Theme.of(context).primaryColor
                                : null,
                            child: Text(
                              items[index],
                              style: TextStyle(fontSize: 22),
                            ),
                          ),
                        ),
                      );
                    }),
              ),
              Wrap(
                children: <Widget>[
                  RawMaterialButton(
                    child: Text('Add rand'),
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    fillColor: Colors.blueGrey,
                    onPressed: () {
                      setState(() {
                        final randItem =
                            'RandItem' + random.nextInt(256).toString();

                        final randomIndex =
                            items.isEmpty ? 0 : random.nextInt(items.length);
                        items.insert(randomIndex, randItem);
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Remove rand'),
                    fillColor: Colors.lightGreen,
                    onPressed: () {
                      setState(() {
                        if (items.length == 1) {
                          items.removeAt(0);
                        } else {
                          items.removeAt(random.nextInt(items.length - 1));
                        }
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    fillColor: Colors.blueGrey,
                    child: Text('Delete'),
                    onPressed: () {
                      setState(() {
                        final itemsToRemove = multiselectController
                            .getSelectedItems()
                            .cast<String>();

                        items = items
                            .where(
                                (element) => !itemsToRemove.contains(element))
                            .toList();
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Select 0'),
                    fillColor: Colors.lightGreen,
                    onPressed: () {
                      setState(() {
                        multiselectController.select(0);
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Select all'),
                    fillColor: Colors.amber,
                    onPressed: () {
                      setState(() {
                        multiselectController.selectAll();
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Invert'),
                    fillColor: Colors.tealAccent,
                    onPressed: () {
                      setState(() {
                        multiselectController.invertSelection();
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Clear'),
                    fillColor: Colors.deepPurpleAccent,
                    onPressed: () {
                      setState(() {
                        multiselectController.clearSelection();
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('No keep'),
                    fillColor: Colors.deepPurpleAccent,
                    onPressed: () {
                      setState(() {
                        keepSelectedIndexes = !keepSelectedIndexes;
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Select specified'),
                    fillColor: Colors.deepPurpleAccent,
                    onPressed: () {
                      setState(() {
                        multiselectController.setSelectedIndexes([0, 1, 2]);
                      });
                    },
                  ),
                ],
              )
            ],
          ),
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/multiselect.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

class Multiselect extends StatefulWidget {
  final Widget child;
  Multiselect({Key key, @required this.child}) : super(key: key);

  static MultiselectState of(BuildContext context) {
    return context
        .dependOnInheritedWidgetOfExactType<_InheritedMultiselect>()
        .data;
  }

  @override
  MultiselectState createState() => MultiselectState();
}

class MultiselectState extends State<Multiselect> {
  List<int> _selectedIndexes = [];
  List<int> get selectedIndexes => _selectedIndexes;

  select(int index) {
    if (!_selectedIndexes.contains(index)) {
      //debugPrint("!_selectedIndexes.contains");
      setState(() {
        _selectedIndexes.add(index);
      });
    }
  }

  unselectItem(int index) {
    if (_selectedIndexes.contains(index)) {
      setState(() {
        _selectedIndexes.remove(index);
      });
    }
  }

  bool indexIsSelected(int i) {
    return _selectedIndexes.contains(i);
  }

  @override
  Widget build(BuildContext context) {
    return _InheritedMultiselect(child: widget.child, data: this);
  }
}

class _InheritedMultiselect extends InheritedWidget {
  _InheritedMultiselect({this.data, Key key, this.child})
      : super(key: key, child: child);

  final Widget child;
  final MultiselectState data;

  // static _InheritedMultiselect of(BuildContext context) {
  //   return context.dependOnInheritedWidgetOfExactType<_InheritedMultiselect>();
  // }

  @override
  bool updateShouldNotify(_InheritedMultiselect oldWidget) {
    return true; //oldWidget.data != data;
  }
}


lib/multiselect_scope.dart
library multiselect_scope;

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

enum SelectionEvent {
  /// Unselect item if it selected, select otherwise
  auto,

  /// Select item
  select,

  /// Deselect item
  unselect,
}

/// An object that stores the selected indexes and also allows you to change them
class MultiselectController extends ChangeNotifier {
  List<int> _selectedIndexes = [];
  List _dataSource = [];

  List<int> get selectedIndexes => _selectedIndexes;
  bool get selectionAttached => _selectedIndexes.any((element) => true);

  late int _itemsCount;

  /// Select (or unselect) item by index
  /// If passed [event] is [SelectionEvent.auto] function automatically
  /// decide that action will need apply
  void select(int index, {SelectionEvent event = SelectionEvent.auto}) {
    final indexContains = _selectedIndexes.contains(index);
    final computedEvent = event == SelectionEvent.auto
        ? indexContains
            ? SelectionEvent.unselect
            : SelectionEvent.select
        : event;

    if (computedEvent == SelectionEvent.select) {
      if (!indexContains) {
        _selectedIndexes.add(index);
        notifyListeners();
      }
    } else if (computedEvent == SelectionEvent.unselect) {
      if (indexContains) {
        _selectedIndexes.remove(index);
        notifyListeners();
      }
    }
  }

  /// Get current selected items in [dataSource]
  List getSelectedItems() {
    final selectedItems = selectedIndexes.map((e) => _dataSource[e]).toList();

    return selectedItems;
  }

  /// Set all selection to empty
  void clearSelection() {
    if (selectedIndexes.any((element) => true)) {
      selectedIndexes.clear();
      notifyListeners();
    }
  }

  /// Replace selection by all not selected items
  void invertSelection() {
    _selectedIndexes = List<int>.generate(_itemsCount, (i) => i)
        .toSet()
        .difference(_selectedIndexes.toSet())
        .toList();

    notifyListeners();
  }

  /// Select all items in [dataSource]
  void selectAll() {
    _selectedIndexes = List<int>.generate(_itemsCount, (i) => i);
    notifyListeners();
  }

  /// Check selection of item by it index
  bool isSelected(int index) {
    return _selectedIndexes.contains(index);
  }

  /// Set selection by specified indexes
  /// Replace existing selected indexes by [newIndexes]
  void setSelectedIndexes(List<int> newIndexes) {
    _setSelectedIndexes(newIndexes, true);
  }

  void _setDataSource(List dataSource) {
    _dataSource = dataSource;
    _itemsCount = dataSource.length;
  }

  void _setSelectedIndexes(List<int> newIndexes, bool notifyListeners) {
    _selectedIndexes = newIndexes;

    if (notifyListeners) {
      this.notifyListeners();
    }
  }
}

typedef SelectionChangedCallback<T> = void Function(
    List<int> selectedIndexes, List<T> selectedItems);

/// Widget to manage item selection
class MultiselectScope<T> extends StatefulWidget {
  /// A child widget that usually contains in its subtree a list
  /// of items whose selection you want to control
  final Widget child;

  /// Function that invoked when selected indexes changes.
  /// Builds appropriate listeners on stage of init [MultiselectScope] widget
  /// and then does not change.
  /// This function will not invoke on first load of this widget.
  final SelectionChangedCallback<T>? onSelectionChanged;

  /// An object that stores the selected indexes and also allows you to change them
  /// This object may be set once and can not be replaced
  /// when updating the widget configuration
  final MultiselectController? controller;

  /// Data for selection tracking
  /// For example list of `Cars` or `Employes`
  final List<T> dataSource;

  /// Clear selection if user push back button
  final bool clearSelectionOnPop;

  /// If [true]: when you update [dataSource] then selected indexes will update
  /// so that the same elements in new [dataSource] are selected
  /// If [false]: selected indexes will have not automatically updates during [dataSource] update
  final bool keepSelectedItemsBetweenUpdates;

  /// Selected indexes, which will be initialized
  /// when the widget is inserted into the widget tree
  final List<int>? initialSelectedIndexes;

  MultiselectScope({
    Key? key,
    required this.dataSource,
    this.controller,
    this.onSelectionChanged,
    this.clearSelectionOnPop = false,
    this.keepSelectedItemsBetweenUpdates = true,
    this.initialSelectedIndexes,
    required this.child,
  }) : super(key: key);

  @override
  _MultiselectScopeState<T> createState() => _MultiselectScopeState<T>();

  static MultiselectController controllerOf(BuildContext context) {
    return context
        .dependOnInheritedWidgetOfExactType<_InheritedMultiselectNotifier>()!
        .controller;
  }
}

class _MultiselectScopeState<T> extends State<MultiselectScope<T>> {
  late List<int> _hashesCopy;
  late MultiselectController _multiselectController;

  void _onSelectionChangedFunc() {
    if (widget.onSelectionChanged != null) {
      widget.onSelectionChanged!(_multiselectController.selectedIndexes,
          _multiselectController.getSelectedItems().cast<T>());
    }
  }

  List<int> _createHashesCopy(MultiselectScope widget) {
    return widget.dataSource.map((e) => e.hashCode).toList();
  }

  @override
  void initState() {
    super.initState();

    _multiselectController = widget.controller ?? MultiselectController();

    _hashesCopy = _createHashesCopy(widget);
    _multiselectController._setDataSource(widget.dataSource);

    if (widget.initialSelectedIndexes != null) {
      _multiselectController._setSelectedIndexes(
          widget.initialSelectedIndexes!, false);
    }

    if (widget.onSelectionChanged != null) {
      _multiselectController.addListener(_onSelectionChangedFunc);
    }
  }

  @override
  void dispose() {
    super.dispose();

    _multiselectController.removeListener(_onSelectionChangedFunc);
    _multiselectController.dispose();
  }

  @override
  void didUpdateWidget(MultiselectScope<T> oldWidget) {
    super.didUpdateWidget(oldWidget);
    debugPrint('didUpdateWidget GreatMultiselect');

    if (widget.keepSelectedItemsBetweenUpdates) {
      _updateController(oldWidget);
    }

    _multiselectController._setDataSource(widget.dataSource);
  }

  @override
  Widget build(BuildContext context) {
    debugPrint('build GreatMultiselect');
    return widget.clearSelectionOnPop
        ? WillPopScope(
            onWillPop: () async {
              if (_multiselectController.selectionAttached) {
                _multiselectController.clearSelection();
                return false;
              }

              return true;
            },
            child: _buildMultiselectScope(),
          )
        : _buildMultiselectScope();
  }

  _InheritedMultiselectNotifier _buildMultiselectScope() =>
      _InheritedMultiselectNotifier(
          child: widget.child, controller: _multiselectController);

  void _updateController(MultiselectScope oldWidget) {
    if (!oldWidget.keepSelectedItemsBetweenUpdates &&
        widget.keepSelectedItemsBetweenUpdates) {
      // Recalculate hashes of previous state
      _hashesCopy = _createHashesCopy(oldWidget);
    }

    final newHashesCopy = _createHashesCopy(widget);

    //debugPrint(
    //    "Old dataSource: ${_hashesCopy} new dataSource: ${newHashesCopy}");
    final oldSelectedHashes = _multiselectController.selectedIndexes
        .map((e) => _hashesCopy[e])
        .toList();

    final newIndexes = <int>[];
    newHashesCopy.asMap().forEach((index, value) {
      //debugPrint("$index $value");
      if (oldSelectedHashes.contains(value)) {
        newIndexes.add(index);
      }
    });

    _multiselectController._setSelectedIndexes(newIndexes, false);
    _hashesCopy = newHashesCopy;
  }
}

class _InheritedMultiselectNotifier
    extends InheritedNotifier<MultiselectController> {
  final MultiselectController controller;

  const _InheritedMultiselectNotifier(
      {Key? key, required Widget child, required this.controller})
      : super(key: key, child: child, notifier: controller);
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:multiselect_scope/multiselect_scope.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Multiselect Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Multiselect'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);

  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  late List<String> _items;
  late MultiselectController _multiselectController;
  late Random random;

  @override
  void initState() {
    super.initState();
    random = Random();
    _items = List.generate(10, (index) => 'Item $index');

    _multiselectController = MultiselectController();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: MultiselectScope<String>(
        controller: _multiselectController,
        dataSource: _items,
        clearSelectionOnPop: true,
        keepSelectedItemsBetweenUpdates: true,
        initialSelectedIndexes: [1, 3],
        onSelectionChanged: (indexes, items) {
          debugPrint(
              'Custom listener invoked! Indexes: $indexes Items: $items');
          return;
        },
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.start,
            children: <Widget>[
              Expanded(
                child: ListView.builder(
                    itemCount: _items.length,
                    itemBuilder: (context, index) {
                      final controller = MultiselectScope.controllerOf(context);

                      final itemIsSelected = controller.isSelected(index);

                      return InkWell(
                        onLongPress: () {
                          if (!controller.selectionAttached) {
                            controller.select(index);
                          }
                        },
                        onTap: () {
                          debugPrint('Item is selected: $itemIsSelected');

                          if (controller.selectionAttached) {
                            controller.select(index);
                          }
                        },
                        child: Padding(
                          padding: const EdgeInsets.all(8.0),
                          child: Container(
                            color: itemIsSelected
                                ? Theme.of(context).primaryColor
                                : null,
                            child: Text(
                              _items[index],
                              style: TextStyle(fontSize: 22),
                            ),
                          ),
                        ),
                      );
                    }),
              ),
              Wrap(
                children: <Widget>[
                  RawMaterialButton(
                    child: Text('Add rand'),
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    fillColor: Colors.blueGrey,
                    onPressed: () {
                      setState(() {
                        final randItem =
                            'RandItem' + random.nextInt(256).toString();

                        final randomIndex =
                            _items.isEmpty ? 0 : random.nextInt(_items.length);
                        _items.insert(randomIndex, randItem);
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Remove rand'),
                    fillColor: Colors.lightGreen,
                    onPressed: () {
                      setState(() {
                        if (_items.length == 1) {
                          _items.removeAt(0);
                        } else {
                          _items.removeAt(random.nextInt(_items.length - 1));
                        }
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    fillColor: Colors.blueGrey,
                    child: Text('Delete'),
                    onPressed: () {
                      setState(() {
                        final itemsToRemove = _multiselectController
                            .getSelectedItems()
                            .cast<String>();

                        _items = _items
                            .where(
                                (element) => !itemsToRemove.contains(element))
                            .toList();
                      });
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Select 0'),
                    fillColor: Colors.lightGreen,
                    onPressed: () {
                      _multiselectController.select(0);
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Select all'),
                    fillColor: Colors.amber,
                    onPressed: () {
                      _multiselectController.selectAll();
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Invert'),
                    fillColor: Colors.tealAccent,
                    onPressed: () {
                      _multiselectController.invertSelection();
                    },
                  ),
                  RawMaterialButton(
                    padding: EdgeInsets.symmetric(horizontal: 8.0),
                    child: Text('Clear'),
                    fillColor: Colors.deepPurpleAccent,
                    onPressed: () {
                      _multiselectController.clearSelection();
                    },
                  ),
                ],
              )
            ],
          ),
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
